node/lib/util/commit.js

```diff
@@ -2222,3 +2222,99 @@ it empty. You can remove the commit entirely with "git meta reset HEAD^".`);
                                        message,
                                        subMessages);
 });
+
+const submoduleHasChangesToCommit = function(status, submoduleName) {
+    const submodule = status.submodules[submoduleName];
+    if (!submodule) {
+        return false;
+    }
+    const { workdir } = submodule;
+    return null !== workdir && !workdir.status.isIndexClean();
+};
+
+const submoduleHasPreCommitHook = co.wrap(function *(repo, submoduleName) {
+    const Hook          = require("../util/hook");
+    const SubmoduleUtil = require("../util/submodule_util");
+
+    const subRepo = yield SubmoduleUtil.getRepo(repo, submoduleName);
+    return Hook.hasHook(subRepo, "pre-commit");
+});
+
+const filterInParallel = co.wrap(function *(collection, asyncPredicate) {
+    const DoWorkQueue = require("../util/do_work_queue");
+
+    const booleans = yield DoWorkQueue.doInParallel(collection,
+        asyncPredicate);
+    return collection.filter((item, index) => booleans[index]);
+});
+
+const getSubmodulesNameForPreCommitCheck = co.wrap(function *(repo,
```

Comment:
1. Rename to "getSubmoduleNamesForPrecommitCheck" (note that the "s" moved)

2. You could slightly simplify the code by having this return the submodule repos instead of the names, since you're going to need them later to execute the hooks.

Reply:
1. renamed.
2. I think returning strings make interface (return type) more flexible for client code to consume and also it is easier to stub during unit tests. Please let me know if you feel strongly that we should change this to return repos.